### PR TITLE
Neutrino filter support changes.

### PR DESCRIPTION
**Pull request by pmienk**

### CODE REVIEW

File: src/interface/block_chain.cpp [hunk 1]
```diff
@@ -145,6 +145,116 @@ bool block_chain::get_header(chain::header& out_header, size_t& out_height,
...
 }
 
+bool block_chain::get_filter(system::data_chunk& out_filter,
+    system::hash_digest& out_hash, size_t height, uint8_t filter_type,
+    bool candidate) const
+{
+    if ((!settings_.bip158) || (filter_type != bc::neutrino_filter_type))
+        return false;
+
+    auto result = database_.blocks().get(height, candidate);
```
Comment (evoskuil) [inline src/interface/block_chain.cpp]: use const

File: src/interface/block_chain.cpp [hunk 2]
```diff
@@ -145,6 +145,116 @@ bool block_chain::get_header(chain::header& out_header, size_t& out_height,
...
+
+    if (!result)
+        return false;
+
+    auto result_filter = database_.neutrino_filters().get(
```
Comment (evoskuil) [inline src/interface/block_chain.cpp]: use const

File: src/interface/block_chain.cpp [hunk 3]
```diff
@@ -145,6 +145,116 @@ bool block_chain::get_header(chain::header& out_header, size_t& out_height,
...
+}
+
+bool block_chain::get_filter(system::data_chunk& out_filter, size_t& out_height,
+    const system::hash_digest& block_hash, uint8_t filter_type,
+    bool candidate) const
+{
+    if ((!settings_.bip158) || (filter_type != bc::neutrino_filter_type))
+        return false;
+
+    auto result = database_.blocks().get(block_hash);
```
Comment (evoskuil) [inline src/interface/block_chain.cpp]: use const

File: src/interface/block_chain.cpp [hunk 4]
```diff
@@ -145,6 +145,116 @@ bool block_chain::get_header(chain::header& out_header, size_t& out_height,
...
+    if (!result)
+        return false;
+
+    const auto height = result.height();
+
+    // The only way to know if a header is indexed is from its presence in the
+    // index. It will not be marked as a candidate until validated as such.
+    if (database_.blocks().get(height, candidate))
+    {
+        auto result_filter = database_.neutrino_filters().get(
```
Comment (evoskuil) [inline src/interface/block_chain.cpp]: use const

File: src/interface/block_chain.cpp [hunk 5]
```diff
@@ -145,6 +145,116 @@ bool block_chain::get_header(chain::header& out_header, size_t& out_height,
...
+}
+
+bool block_chain::get_filter_header(system::hash_digest& out_filter_header,
+    system::hash_digest& out_hash, size_t height, uint8_t filter_type,
+    bool candidate) const
+{
+    if ((!settings_.bip158) || (filter_type != bc::neutrino_filter_type))
+        return false;
+
+    auto result = database_.blocks().get(height, candidate);
```
Comment (evoskuil) [inline src/interface/block_chain.cpp]: use const

File: src/interface/block_chain.cpp [hunk 6]
```diff
@@ -145,6 +145,116 @@ bool block_chain::get_header(chain::header& out_header, size_t& out_height,
...
+
+    if (!result)
+        return false;
+
+    auto result_filter = database_.neutrino_filters().get(
```
Comment (evoskuil) [inline src/interface/block_chain.cpp]: use const

File: src/interface/block_chain.cpp [hunk 7]
```diff
@@ -145,6 +145,116 @@ bool block_chain::get_header(chain::header& out_header, size_t& out_height,
...
+}
+
+bool block_chain::get_filter_header(system::hash_digest& out_filter_header,
+    size_t& out_height, const system::hash_digest& block_hash,
+    uint8_t filter_type, bool candidate) const
+{
+    if ((!settings_.bip158) || (filter_type != bc::neutrino_filter_type))
+        return false;
+
+    auto result = database_.blocks().get(block_hash);
```
Comment (evoskuil) [inline src/interface/block_chain.cpp]: use const

File: src/interface/block_chain.cpp [hunk 8]
```diff
@@ -145,6 +145,116 @@ bool block_chain::get_header(chain::header& out_header, size_t& out_height,
...
+    if (!result)
+        return false;
+
+    const auto height = result.height();
+
+    // The only way to know if a header is indexed is from its presence in the
+    // index. It will not be marked as a candidate until validated as such.
+    if (database_.blocks().get(height, candidate))
+    {
+        auto result_filter = database_.neutrino_filters().get(
```
Comment (evoskuil) [inline src/interface/block_chain.cpp]: use const

File: src/interface/block_chain.cpp [hunk 9]
```diff
@@ -279,6 +389,25 @@ void block_chain::populate_pool_transaction(const chain::transaction& tx,
...
+    if (!settings_.bip158)
+        return;
+
+    const auto& header = block.header();
+    auto previous_block = database_.blocks().get(header.previous_block_hash());
+    auto previous_filter = database_.neutrino_filters().get(
+        previous_block.neutrino_filter());
+
+    auto filter = system::neutrino::compute_filter(block);
+    auto filter_header = system::neutrino::compute_filter_header(
```
Comment (evoskuil) [inline src/interface/block_chain.cpp]: use const ( 4 lines)

File: src/interface/block_chain.cpp (hunk 1 above)
Comment (pmienk) [reply src/interface/block_chain.cpp]: Corrected.

File: src/interface/block_chain.cpp (hunk 3 above)
Comment (pmienk) [reply src/interface/block_chain.cpp]: Corrected.

File: src/interface/block_chain.cpp (hunk 4 above)
Comment (pmienk) [reply src/interface/block_chain.cpp]: Corrected.

File: src/interface/block_chain.cpp (hunk 5 above)
Comment (pmienk) [reply src/interface/block_chain.cpp]: Corrected.

File: src/interface/block_chain.cpp (hunk 7 above)
Comment (pmienk) [reply src/interface/block_chain.cpp]: Corrected.

File: src/interface/block_chain.cpp (hunk 9 above)
Comment (pmienk) [reply src/interface/block_chain.cpp]: Corrected.

File: src/interface/block_chain.cpp (hunk 8 above)
Comment (pmienk) [reply src/interface/block_chain.cpp]: Corrected.

File: src/interface/block_chain.cpp (hunk 6 above)
Comment (pmienk) [reply src/interface/block_chain.cpp]: Corrected.

File: src/interface/block_chain.cpp (hunk 2 above)
Comment (pmienk) [reply src/interface/block_chain.cpp]: Corrected.

**Pull request merged**